// @moralisweb3/streams/lib/resolvers/createStreamEvm.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStreamEvm = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var name = 'CreateStream';
var method = 'put';
var bodyParams = [
    'webhookUrl',
    'description',
    'tag',
    'tokenAddress',
    'topic0',
    'includeNativeTxs',
    'filter',
    'address',
    'type',
    'chainIds',
    'abi',
];
var apiToResult = function (apiData) {
    var data = (0, core_1.toCamelCase)(apiData);
    return __assign(__assign({}, data), { address: data.address ? evm_utils_1.EvmAddress.create(data.address) : undefined, tokenAddress: data.tokenAddress ? evm_utils_1.EvmAddress.create(data.tokenAddress) : undefined, chains: data.chainIds.map(function (chainId) { return evm_utils_1.EvmChain.create(chainId); }) });
};
exports.createStreamEvm = (0, api_utils_1.createEndpointFactory)(function () {
    return (0, api_utils_1.createEndpoint)({
        name: name,
        getUrl: function () { return "/streams/evm"; },
        apiToResult: apiToResult,
        resultToJson: function (data) { return (__assign(__assign({}, data), { address: data.address ? data.address.format() : undefined, tokenAddress: data.tokenAddress ? data.tokenAddress.format() : undefined, chainIds: data.chains.map(function (chain) { return chain.format(); }) })); },
        parseParams: function (params) { return (__assign(__assign({}, params), { tokenAddress: params.tokenAddress ? evm_utils_1.EvmAddress.create(params.tokenAddress).checksum : undefined, address: params.address ? evm_utils_1.EvmAddress.create(params.address).checksum : undefined, chainIds: params.chains.map(function (chain) { return evm_utils_1.EvmChain.create(chain).apiHex; }) })); },
        method: method,
        bodyParams: bodyParams,
    });
});
