// @moralisweb3/evm-api/lib/resolvers/token/getTokenTransfers.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTokenTransfers = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
// TODO: add decimals from api
exports.getTokenTransfers = (0, api_utils_1.createPaginatedEndpointFactory)(function (core) {
    return (0, api_utils_1.createPaginatedEndpoint)({
        name: 'getTokenTransfers',
        urlParams: ['address'],
        getUrl: function (params) { return "/erc20/".concat(params.address, "/transfers"); },
        apiToResult: function (data, params) {
            var _a;
            return ((_a = data.result) !== null && _a !== void 0 ? _a : []).map(function (transfer) {
                return evm_utils_1.Erc20Transfer.create(__assign(__assign({}, (0, core_1.toCamelCase)(transfer)), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core), address: evm_utils_1.EvmAddress.create(transfer.address, core), toAddress: evm_utils_1.EvmAddress.create(transfer.to_address, core), fromAddress: evm_utils_1.EvmAddress.create(transfer.from_address, core), value: core_1.BigNumber.create(transfer.value), blockTimestamp: new Date(transfer.block_timestamp) }));
            });
        },
        resultToJson: function (data) { return data.map(function (transfer) { return transfer.toJSON(); }); },
        parseParams: function (params) { return (__assign(__assign({}, params), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex, address: evm_utils_1.EvmAddress.create(params.address, core).lowercase, to_block: params.toBlock, from_block: params.fromBlock, from_date: params.fromDate, to_date: params.toDate })); },
        firstPageIndex: 0,
    });
});
