// @moralisweb3/evm-api/lib/resolvers/balance/getNativeBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNativeBalance = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getNativeBalance = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'getNativeBalance',
        urlParams: ['address'],
        getUrl: function (params) { return "/".concat(params.address, "/balance"); },
        apiToResult: function (data) { return ({
            balance: evm_utils_1.EvmNative.create(data.balance, 'wei'),
        }); },
        resultToJson: function (data) { return ({
            balance: data.balance.format(),
        }); },
        parseParams: function (params) { return ({
            chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex,
            address: evm_utils_1.EvmAddress.create(params.address, core).lowercase,
            to_block: params.toBlock,
            providerUrl: params.providerUrl,
        }); },
    });
});
