// @moralisweb3/evm-api/lib/resolvers/block/getDateToBlock.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDateToBlock = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getDateToBlock = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'getDateToBlock',
        getUrl: function () { return "/dateToBlock"; },
        apiToResult: function (data) { return (__assign(__assign({}, data), { date: new Date(data.date) })); },
        resultToJson: function (data) { return (__assign(__assign({}, data), { date: (0, core_1.dateInputToDate)(data.date) })); },
        parseParams: function (params) { return ({
            chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex,
            date: params.date,
        }); },
    });
});
