// @moralisweb3/auth/lib/resolvers/solVerifyChallenge.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.completeChallengeSol = void 0;
var sol_utils_1 = require("@moralisweb3/sol-utils");
var core_1 = require("@moralisweb3/core");
var api_utils_1 = require("@moralisweb3/api-utils");
var name = 'verifyChallengeSolana';
var method = 'post';
var bodyParams = ['message', 'signature'];
exports.completeChallengeSol = (0, api_utils_1.createEndpointFactory)(function () {
    return (0, api_utils_1.createEndpoint)({
        name: 'Verify Challenge (Solana)',
        getUrl: function () { return "/challenge/verify/solana"; },
        apiToResult: function (_a) {
            var network = _a.network, data = __rest(_a, ["network"]);
            return (__assign(__assign({}, data), { solNetwork: sol_utils_1.SolNetwork.create(network), address: sol_utils_1.SolAddress.create(data.address), expirationTime: (0, core_1.maybe)(data.expirationTime, function (value) { return new Date(value); }) }));
        },
        resultToJson: function (result) { return (__assign(__assign({}, (0, core_1.toCamelCase)(result)), { solNetwork: result.solNetwork.format(), address: result.address.format() })); },
        parseParams: function (params) { return params; },
        method: method,
        bodyParams: bodyParams,
    });
});
