// @moralisweb3/evm-api/lib/resolvers/nft/getContractNFTs.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractNFTs = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getContractNFTs = (0, api_utils_1.createPaginatedEndpointFactory)(function (core) {
    return (0, api_utils_1.createPaginatedEndpoint)({
        name: 'getContractNFTs',
        urlParams: ['address'],
        getUrl: function (params) { return "/nft/".concat(params.address); },
        apiToResult: function (data, params) {
            var _a;
            return ((_a = data.result) !== null && _a !== void 0 ? _a : []).map(function (nft) {
                return evm_utils_1.EvmNft.create(__assign(__assign({}, (0, core_1.toCamelCase)(nft)), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core), ownerOf: nft.owner_of ? evm_utils_1.EvmAddress.create(nft.owner_of, core) : undefined, lastMetadataSync: nft.last_metadata_sync ? new Date(nft.last_metadata_sync) : undefined, lastTokenUriSync: nft.last_token_uri_sync ? new Date(nft.last_token_uri_sync) : undefined }), core);
            });
        },
        resultToJson: function (data) { return data.map(function (nft) { return nft.toJSON(); }); },
        parseParams: function (params) { return (__assign(__assign({}, params), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex, address: evm_utils_1.EvmAddress.create(params.address, core).lowercase })); },
        firstPageIndex: 0,
    });
});
