// @moralisweb3/evm-api/lib/resolvers/nft/getNFTLowestPrice.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNFTLowestPrice = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getNFTLowestPrice = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'getNFTLowestPrice',
        urlParams: ['address'],
        getUrl: function (params) { return "/nft/".concat(params.address, "/lowestprice"); },
        apiToResult: function (data, params) {
            return evm_utils_1.EvmNftTrade.create(__assign(__assign({}, (0, core_1.toCamelCase)(data)), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core), sellerAddress: evm_utils_1.EvmAddress.create(data.seller_address, core), buyerAddress: evm_utils_1.EvmAddress.create(data.buyer_address, core), marketplaceAddress: evm_utils_1.EvmAddress.create(data.marketplace_address, core), tokenAddress: evm_utils_1.EvmAddress.create(data.token_address, core), price: evm_utils_1.EvmNative.create(data.price, 'wei'), blockTimestamp: new Date(data.block_timestamp), tokenIds: data.token_ids }));
        },
        resultToJson: function (data) { return data.toJSON(); },
        parseParams: function (params) { return (__assign(__assign({}, params), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex, address: evm_utils_1.EvmAddress.create(params.address, core).lowercase, provider_url: params.providerUrl })); },
    });
});
