// @moralisweb3/evm-api/lib/resolvers/nft/getWalletNFTs.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWalletNFTs = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getWalletNFTs = (0, api_utils_1.createPaginatedEndpointFactory)(function (core) {
    return (0, api_utils_1.createPaginatedEndpoint)({
        name: 'getWalletNFTs',
        urlParams: ['address'],
        getUrl: function (params) { return "/".concat(params.address, "/nft"); },
        apiToResult: function (data, params) {
            var _a;
            return ((_a = data.result) !== null && _a !== void 0 ? _a : []).map(function (nft) {
                return evm_utils_1.EvmNft.create({
                    chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core),
                    contractType: nft.contract_type,
                    tokenAddress: nft.token_address,
                    tokenId: nft.token_id,
                    tokenUri: nft.token_uri,
                    metadata: nft.metadata,
                    name: nft.name,
                    symbol: nft.symbol,
                    amount: nft.amount ? parseInt(nft.amount, 10) : undefined,
                    blockNumberMinted: nft.block_number_minted,
                    blockNumber: nft.block_number,
                    ownerOf: evm_utils_1.EvmAddress.create(nft.owner_of, core),
                    tokenHash: nft.token_hash,
                    lastMetadataSync: (0, core_1.dateInputToDate)(nft.last_metadata_sync),
                    lastTokenUriSync: (0, core_1.dateInputToDate)(nft.last_token_uri_sync),
                });
            });
        },
        resultToJson: function (data) { return data.map(function (nft) { return nft.toJSON(); }); },
        parseParams: function (params) {
            var _a;
            return (__assign(__assign({}, params), { chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex, token_addresses: (_a = params.tokenAddresses) === null || _a === void 0 ? void 0 : _a.map(function (address) { return evm_utils_1.EvmAddress.create(address, core).lowercase; }), address: evm_utils_1.EvmAddress.create(params.address, core).lowercase }));
        },
    });
});
