// @moralisweb3/evm-api/lib/resolvers/nft/syncNFTContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncNFTContract = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
var method = 'put';
exports.syncNFTContract = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'syncNFTContract',
        urlParams: ['address'],
        getUrl: function (params) { return "/nft/".concat(params.address, "/sync"); },
        apiToResult: function (_) { return ({
            success: true,
        }); },
        resultToJson: function () { return ({
            success: true,
        }); },
        parseParams: function (params) { return ({
            chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex,
            address: evm_utils_1.EvmAddress.create(params.address, core).lowercase,
        }); },
        method: method,
    });
});
