// @moralisweb3/evm-api/lib/resolvers/defi/getPairReserves.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPairReserves = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var evm_utils_1 = require("@moralisweb3/evm-utils");
var EvmChainResolver_1 = require("../EvmChainResolver");
exports.getPairReserves = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'getPairReserves',
        urlParams: ['pairAddress'],
        getUrl: function (params) { return "/".concat(params.pairAddress, "/reserves"); },
        apiToResult: function (data) { return data; },
        resultToJson: function (data) { return data; },
        parseParams: function (params) { return ({
            chain: EvmChainResolver_1.EvmChainResolver.resolve(params.chain, core).apiHex,
            pair_address: evm_utils_1.EvmAddress.create(params.pairAddress, core).lowercase,
            provider_url: params.providerUrl,
            to_block: params.toBlock,
            to_date: params.toDate,
        }); },
    });
});
